File: src/main/kotlin/dragonbook/symbols/Env.java
package dragonbook.symbols;

import dragonbook.inter.expr.Id;
import dragonbook.lexer.Token;

import java.util.HashMap;

public class Env {
    private final HashMap<Token, Id> table = new HashMap<>();
    private final Env prev;

    public Env(Env prev) {
        this.prev = prev;
    }

    public void put(Token tok, Id id) {
        table.put(tok, id);
    }

    public Id get(Token tok) {
        for (Env env = this; env != null; env = env.prev) {
            Id found = env.table.get(tok);
            if (found != null) return found;
        }
        return null;
    }
}

File: src/main/kotlin/dragonbook/symbols/Env.kt
package dragonbook.symbols

import dragonbook.inter.expr.Id
import dragonbook.lexer.Token

class Env(private val prev: Env?) {
    private val table = HashMap<Token, Id>()

    operator fun set(tok: Token, id: Id) {
        table[tok] = id
    }

    operator fun get(tok: Token): Id? {
        var env: Env? = this
        while (env != null) {
            val found = env.table[tok]
            if (found != null) return found
            env = env.prev
        }
        return null
    }
}